Adding `NormalizeForStreamProcessor`

eyalkoren opened the pull request:
Namespacing algorithm [EDIT 1]:
1. start by checking whether the document is OTel or not. A document is considered OTel if:
   - `resource` exists as a key and the value is a map
   - `resource` either doesn't contain an `attributes` field, or contains an `attributes` field of type map
   - `scope` is either missing or a map
   - `attributes` is either missing or a map
   - `body` is either missing or a map
   - `body` either doesn't contain a `text` field, or contains a `text` field of type `String` 
   - `body` either doesn't contain a `structured` field, or contains a `structured` field that is not of type `String`
2. if it is OTel - return as is
3. if it is not OTel:
   - create a new `attributes` map   
   - create new `resource` map with one entry of which `attributes` is the key and a new map as its value
   - move the following top level fields (if they exist) to the new `attributes` map: `attributes`, `resource`, `span_id`, `body`, `severity_text` and `trace_id`
   - add the new `attributes` and `resource` maps as top level fields
   - rename special keys (e.g. `span.id`, `log.level`) to OTel-compliant names: for each, look for a value first in the nested form and if not found look for a top level dotted field. The first value that is found is used for the renamed field
   -  move all remaining top level fields, other than `@timestamp`, `trace_id`, `span_id`, `severity_text`, `body`, `attributes`, `resource` and `scope` to the new `attributes` map
   - flatten all fields that are not arrays in `attributes`
   - move specific attributes that describe resources from `attributes` to `resource.attributes`

eyalkoren committed:
Adding EcsNamespacingProcessor

eyalkoren committed:
Adding module-info

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Exposing and testing the processor

eyalkoren committed:
Add test and some algorithm fixes

eyalkoren committed:
Making scope non-mandatory

eyalkoren committed:
Minimize dependencies

eyalkoren committed:
Extending REST tests

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Update docs/changelog/125699.yaml

elasticsearchmachine commented:
Hi @eyalkoren, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

dakrone commented on modules/ingest-ecs/src/test/java/org/elasticsearch/ingest/ecs/EcsNamespacingProcessorTests.java:
```diff
@@ -0,0 +1,403 @@
+        assertTrue(processor.isOTelDocument(document));
+    }
+
+    public void testIsOTelDocument_validOTelDocumentWithScopAndAttributes() {
```
Super minor nit
```suggestion
    public void testIsOTelDocument_validOTelDocumentWithScopeAndAttributes() {
```

dakrone commented on modules/ingest-ecs/src/main/java/org/elasticsearch/ingest/ecs/EcsNamespacingProcessor.java:
```diff
@@ -0,0 +1,217 @@
+
+public class EcsNamespacingProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "ecs_namespacing";
```
Actually, I'm not really sure that `namespace` benefits a whole lot here. What if we called it the `ecs_to_otel` processor, or something similar?

eyalkoren replied:
I'm am not sure this qualifies as a conversion to OTel, I think it is more transformation of ECS to namespaced-ECS, mostly of attributes/resource-attributes, to make it more compatible with OTel queries. @felixbarny WDYT?

eyalkoren replied:
I renamed everything to `ecs_namespace` for now

eyalkoren replied:
Everything would be refactored to `normalize_to_otel`/`NormalizeToOTel` and the module will be renamed to `ingest-otel`

dakrone reviewed (changes requested):
Thanks Eyal, I left some initial comments. I had a question about the way that we nest a document with an existing `attributes` field into the OTel attributes. Is this something we want to do? For example this doc:

```json
{
  "attributes": {
    "a": "b",
    "c": [1, 2, 3]
  },
  "log.level": "1234"
}
```

Becomes, after processing:

```json
{
  "resource": {
    "attributes": {}
  },
  "severity_text": "1234",
  "attributes": {
    "attributes.a": "b",
    "attributes.c": [1, 2, 3]
  }
}
```

Is that the desired behavior for an existing `attributes` field?

eyalkoren commented on modules/ingest-ecs/src/main/java/org/elasticsearch/ingest/ecs/EcsNamespacingProcessor.java:
```diff
@@ -0,0 +1,217 @@
+            if (keepKey.equals("@timestamp")) {
+                continue;
+            }
+            Object value = source.remove(keepKey);
+            if (value != null) {
+                newAttributes.put(keepKey, value);
+            }
```
You are right. It's not equivalent to `containsKey`+`get`. I made the decision that I can make this minor optimization because this key is not going to be present in the resulting document anyway, for instance 
```json
"message" : null
```
would be converted to
```json
{
  "body" : {
     "text" : null
   }
}
```
so if someone was deriving a decision only based on the presence of this key, it would break anyway. However, if we think this is considered a lost of info, I'll restore so that `null` values are represented. Not a big gain in this optimization anyway.

eyalkoren replied:
I made sure to keep/move keys with `null` values

eyalkoren commented on modules/ingest-ecs/src/main/java/org/elasticsearch/ingest/ecs/EcsNamespacingProcessor.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    private boolean shouldMoveToResourceAttributes(String key) {
+        return key.startsWith(AGENT_PREFIX)
+            || key.equals("agent")
+            || key.startsWith(CLOUD_PREFIX)
+            || key.equals("cloud")
+            || key.startsWith(HOST_PREFIX)
+            || key.equals("host");
```
Definitely! I got this reversed at some point and it remained this way, doesn't make sense at all. Thanks!

eyalkoren commented:
Answering the general question:
> I had a question about the way that we nest a document with an existing attributes field into the OTel attributes. Is this something we want to do? 

We started off by trying to be more "clever" about this and merge OTel with non-OTel. Then we had to handle lots of corner cases, like:

- if `attributes` exists and is not a map - it needs to go into a new `attributes` map
- if `resource` exists and is not a map - it needs to go into a new `attributes` map
- if there was `attributes` map before and `resource` is not a map, we need to make sure that the new `attributes.resource` attribute becomes an array that includes both values
- same for `resource.attribtues` 
- if both `span.id` and `span_id` exist - we need to do something about it, for example: make the new `span_id` an array with two values

And so forth.

So last week we decided to change the way we think about it: a document is either sent by an OTel-compliant shipper, or not. If not, no reason to treat the original fields as if they have the OTel sematics. So even if it has a field that has an OTel name, we can consider it to be by chance and namespace it. If that's so- no reason to complicate things for the unlikely event where real OTel documents contain non-OTel fields.

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
instanceOf with pattern matching

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

eyalkoren committed:
instanceOf with pattern matching

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

eyalkoren committed:
revert constants usage

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

eyalkoren committed:
Merge remote-tracking branch 'eyalkoren/ECS-namespacing-processor' into ECS-namespacing-processor

eyalkoren committed:
Complete review change proposals

eyalkoren committed:
fix typo in test name

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

Commit:
[CI] Auto commit changes from spotless

eyalkoren committed:
Applying review suggestions

eyalkoren committed:
Merge remote-tracking branch 'eyalkoren/ECS-namespacing-processor' into ECS-namespacing-processor

joegallo commented:
I've assigned this to myself (to further indicate that it's waiting on my review). I'll review it tomorrow.

joegallo committed:
Use ofEntries for increased clarity

joegallo committed:
Save a rehash and some traversals

joegallo committed:
Rename this variable

joegallo committed:
Silence a warning from Intellij

joegallo committed:
Fix some typos and reflow a comment

joegallo committed:
Merge branch 'main' into ECS-namespacing-processor

joegallo commented on modules/ingest-ecs/src/main/java/org/elasticsearch/ingest/ecs/EcsNamespaceProcessor.java:
```diff
@@ -0,0 +1,275 @@
+
+        // Flatten attributes
+        source.replace(ATTRIBUTES_KEY, Maps.flatten(newAttributes, false, false));
+        newResource.replace(ATTRIBUTES_KEY, Maps.flatten(newResourceAttributes, false, false));
```
`replace` seems like an interesting for these two lines -- why not just `put(...)`? Is there a circumstance in which those keys will not actually be present?

eyalkoren replied:
No, with the current state of the code these keys should always be mapped. I'll change to `put`

joegallo commented:
I added a few relatively trivial commits as a form of review. I'm still getting to know the implementation, so I'll take a closer look at some parts and come back to you if I have any questions or suggestions. Overall it looks pretty good, though.

joegallo committed:
This can be static

joegallo committed:
Rely on mutability for these tests

joegallo committed:
Rename some variables

joegallo committed:
Drop the top-level warnings suppression

joegallo committed:
Merge branch 'main' into ECS-namespacing-processor

joegallo commented on modules/ingest-ecs/src/test/java/org/elasticsearch/ingest/ecs/EcsNamespaceProcessorTests.java:
```diff
@@ -0,0 +1,409 @@
+        Map<String, Object> before = new HashMap<>(source);
+        IngestDocument document = new IngestDocument("index", "id", 1, null, null, source);
+        processor.execute(document);
+        assertEquals(before, document.getSource());
```
I don't think this test is actually testing what its intended to test -- `new HashMap<>(source)` should probably be some kind of deep copy instead.

joegallo replied:
Actually, nevermind, I rewrote it in 65f7ea2a2f613b2a5e3d62d69f9c12880e88cf77.

eyalkoren replied:
It tested what _I intended_ it to test - that keys are not moved from the top level when you process a valid OTel document. A shallow copy is sufficient for this purpose.
In fact, now it doesn't test anything - the source always remains the same instance, you can add this assertion to all tests and they will pass.
I will restore that, keeping your nicer map construction.

joegallo replied:
> In fact, now it doesn't test anything 

I don't think I agree -- since it's an immutable map of immutable maps, the fact that `processor.execute(document)` doesn't throw an exception is probably a suitable proxy for it not having attempted to modify the map.

joegallo replied:
Eh, I'm not going to fight about it. I like it as I wrote it here, but I'll probably be okay with some different version of things, too.

joegallo commented:
When I run `./gradlew ":modules:ingest-ecs:check` for this branch, it fails -- perhaps there's something incorrect in the `build.gradle` for it (or something missing, 🤷)?

joegallo committed:
Rewrite this test

Assert that it's the sameInstance, and use an immutable map of
immutable maps in order to be sure that nothing changed.

joegallo commented on modules/ingest-ecs/src/main/java/org/elasticsearch/ingest/ecs/EcsNamespaceProcessor.java:
```diff
@@ -0,0 +1,275 @@
+            if (keepKey.equals("@timestamp")) {
+                continue;
+            }
+            Object value = source.remove(keepKey);
```
Does it matter that this will remove things that are explicitly mapped to `null` and not re-map them?

eyalkoren replied:
I answered that [above](https://github.com/elastic/elasticsearch/pull/125699#discussion_r2020052854), but I think that even this doubt doesn't worth the very small optimization. I will replace with `containsKey` and maintain keys with `null` values.

eyalkoren commented:
Thanks for the review @joegallo 🙏 
Note that there are few more additions I need to make to this PR:
1. it was decided to add more namespaces to `attribute.resource`, it's just not finalized yet exactly which
2. after adding those - I will update and add tests accordingly
3. I need to rename the processor type to `normalize_to_otel` and the processor class name to `NormalizeToOtelProcessor`
4. add it to documentation

I am waiting for a finalized list of keys that need to go into `resource.attributes` and then I will wrap it up and request a final review, which should be quite trivial.

> When I run ./gradlew ":modules:ingest-ecs:check for this branch, it fails -- perhaps there's something incorrect in the build.gradle for it (or something missing, 🤷)?

gradle is definitely not a strength of mine and Elasticsearch is not the best project to get familiar with its basics 🙂 
If you have any idea what doesn't seem right - let me know. Otherwise, I'll find what it is.

flash1293 commented:
We discussed that we want to avoid making this processor available to users too quickly, so let's not expose it on serverless right away. According to @joegallo it's possible to filter based on the package name to achieve this.

We still plan to make this available eventually, but since discussions around details of the central /logs endpoint are still ongoing and could affect this, it makes sense to not move unnecessary quickly here, since the value of the processor will only be achieved with the combination of logs endpoint and streams API/UI changes anyway.

joegallo commented:
+1 to @flash1293's comment that we're pretty sure we can filter this out from serverless for the time being. We'll probably want to flag that this is a tech preview (or some such) in the ordinary documentation (and that documentation may come via a different PR, I'm not sure what the intent is necessarily).

joegallo committed:
Merge branch 'main' into ECS-namespacing-processor

joegallo committed:
Drop yaml-rest-compat-test

There aren't any tests of this on 8.19, so there's nothing to test
this against in a yaml-rest-compat-test way.

joegallo commented:
> gradle is definitely not a strength of mine and Elasticsearch is not the best project to get familiar with its basics 🙂

Heh, I know **exactly** what you mean. 😄 

I poked and prodded things for a little while this afternoon, though, and I got it to stop being unhappy with us for now at least. I'm going to watch this new build to see if we get to a green check mark from CI -- there was a serverless failure that kept popping up which **hopefully** will go away now. 🤷

eyalkoren committed:
Refactor ECS Namespacing to Normalize to OTel

eyalkoren committed:
Apply review comments

eyalkoren committed:
Adding accurate resource attributes handling

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Suppress warning for forbidden usage of System.out in tests

eyalkoren committed:
Eliminating some more forbidden APIs

Commit:
[CI] Auto commit changes from spotless

eyalkoren committed:
Reverting shameful refactoring errors

eyalkoren committed:
Merge remote-tracking branch 'eyalkoren/ECS-namespacing-processor' into ECS-namespacing-processor

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Disabling ResourceAttributesTests

eyalkoren committed:
Adding local disk crawler

eyalkoren committed:
Disabling tests with @LuceneTestCase.Nightly()

eyalkoren committed:
Disabling forbidden APIs

eyalkoren committed:
Disabling test and adding javadoc

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Fix Logger#warn usage

eyalkoren commented:
Thanks for the feedback @dakrone 🙏 

> I would prefer to avoid a file, and I think programmatic updates should be avoided as well. We have to consider even additions as something we may want to have a discussion about rather than automatically applying them.

There is no intention to automatically accept changes in behavior without a review. The idea is to programmatically create the file when a required change is detected, **_in addition to failing the nightly/weekly check and notifying us_**. This leaves us only to review and move a file in most cases, without any code changes. If we need versioning - this makes it a bit easier as well. Anyway, the test failure will omit the required changes in the set, so it's not going to be a big effort to update either way. We'll stay with a static immutable set.

> We can do this in subsequent work, and in this PR still have the disabled test committed. Then we can work with the delivery team on how to run this periodically.

No problem- I will open a followup issue once we merge this. But what's the right way to disable a test? The build fails with this name-based method and also if I use `@Ignore`, or `@Nightly` or anything else I tried.

eyalkoren committed:
Adding documentation

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren commented:
I added proposed documentation.

@flash1293 @felixbarny please go over the documentation contents

@joegallo @dakrone please go over the documentation general structure. Most importantly - I used a `note` to describe this as tech preview (is this the case?) and I used "serverless", but probably there is a better end-user wording for that.

flash1293 commented on docs/reference/enrich-processor/normalize-to-otel-processor.md:
```diff
@@ -0,0 +1,149 @@
+# Normalize-to-OTel processor [normalize-to-otel-processor]
+
+
+Detects whether a document is OpenTelemetry-compliant and if not - normalizes it as described below. The resulting document can be queried seamlessly by clients that expect either [ECS](https://www.elastic.co/guide/en/ecs/current/index.html) or OpenTelemetry-[Semantic-Conventions](https://github.com/open-telemetry/semantic-conventions) formats.
```
The resulting doc can not automatically be queried both ways, it needs the otel mapping component templates for that as well. We should either add that or leave it out.

eyalkoren replied:
Of course! I will add this. I don't think it makes sense to leave it out as this is the purpose of this processor, right? This one sentence is important, it should capture the essence of what this processor is good for. 

If you want to write something for this summary - I think it would be best. Otherwise, I will propose a version that includes the OTel component templates. Please refer me to which are they.

flash1293 replied:
It's fine to start small. We should reference the `logs-otel@template` index template here.

eyalkoren replied:
@flash1293 please see if the [current wording](https://github.com/elastic/elasticsearch/pull/125699/files#diff-12dc3d32727f60ba943302b2c52725b23ee9b820093713e65c997c625e482927R10-R14) is what you expect

flash1293 replied:
Looks great, thanks Eyal!

flash1293 commented on docs/reference/enrich-processor/normalize-to-otel-processor.md:
```diff
@@ -0,0 +1,149 @@
+    * The corresponding OpenTelemetry attribute is defined in
+      [Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/tree/main/model)
+      within a group that is defined as `type: enitity`.
+* All other fields, except from `@timestamp` are moved to the `attributes` map.
```
except for

flash1293 reviewed (commented):
Looked mostly at the documentation, left some small notes.

eyalkoren committed:
Fix typo

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Disabling the tests temporarily with @Ignore

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Adding to toc

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Remove GitHub-API-based crawler

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ECS-namespacing-processor

eyalkoren committed:
Refactor: renaming to normalize_for_stream

joegallo committed:
Merge branch 'main' into ECS-namespacing-processor

joegallo committed:
Merge branch 'main' into ECS-namespacing-processor

joegallo reviewed (approved):
I opened a ticket for us to track adding the periodic ci to run the otel-semver-crawler test bits, but I'm fine with that staying disabled for now on this PR.

I see also that there's a perhaps unfinished conversation about the wording of the documentation. I'm okay if that's either resolved here and now, or if we fuss with the wording of things in a follow up PR.

✅

joegallo commented:
I believe it's the case that this PR is intended to be backported to 8.19.0, so I'm going to add that label, and also add the label for attempting to automatically backport it. Feel free to remove those labels if this is not actually intended for 8.19.0, though.

dakrone commented:
This should go to 8.19

joegallo committed:
Merge branch 'main' into ECS-namespacing-processor

eyalkoren committed:
Merge branch 'main' into ECS-namespacing-processor

eyalkoren commented:
CI keeps being unhappy with random stuff, I don't think related to this PR contents.
Once we get a green build, are we good to merge?

Regarding the followup task of running the test nightly - do you want me to open a GH issue with summary of what I already discussed with the delivery team?

joegallo commented:
> Once we get a green build, are we good to merge?

For my part, yes. 👍 (And quick, it's green!)
